lib/dataloader/ecto.ex

```diff
@@ -364,18 +364,27 @@ if Code.ensure_loaded?(Ecto) do
       end
 
       def load(source, batch, item) do
-        case fetch(source, batch, item) do
-          {:error, _message} ->
-            batch = normalize_key(batch, source.default_params)
-            {batch_key, item_key, item} = get_keys(batch, item)
-            entry = {item_key, item}
+        {batch_key, item_key, item} =
+          batch
+          |> normalize_key(source.default_params)
+          |> get_keys(item)
 
-            update_in(source.batches, fn batches ->
-              Map.update(batches, batch_key, MapSet.new([entry]), &MapSet.put(&1, entry))
-            end)
+        if fetched?(source.results, batch_key, item_key) do
+          source
+        else
+          entry = {item_key, item}
 
-          _ ->
-            source
+          update_in(source.batches, fn batches ->
+            Map.update(batches, batch_key, MapSet.new([entry]), &MapSet.put(&1, entry))
+          end)
+        end
+      end
+
+      defp fetched?(results, batch_key, item_key) do
+        case Map.fetch(results, batch_key) do
+          :error -> false
+          {:ok, {:error, _reason}} -> false
+          {:ok, {:ok, batch}} -> Map.has_key?(batch, item_key)
```

Comment:
Directly pattern matching on the map instead of `Map.fetch` should provide some performance benefits by avoiding allocating the `ok` tuple. I'm not sure how frequently this is called and if it would have any benefit, though.

Reply:
I'm not sure if you mean this for `Map.has_key?/2` only, or if you also meant this for the `Map.fetch/2` call just before, but I think it applies to that call as well, which allows something like this:

```elixir
      defp fetched?(%{^batch_key => {:ok, %{^item_key => _value}}}, batch_key, item_key), do: true
      defp fetched?(_results, _batch_key, _item_key), do: false
```

Reply:
Unfortunately pattern matching doesn't work that way, you can't have the key specified in one arg and then use that to match in another. You could do so with a `case` though.

Reply:
Yeah, I meant something like:
```elixir
case results do
  %{^batch_key => {:ok, %{^item_key => _}}} -> true
  _ -> false
end
```

Reply:
TIL. Thank you for the feedback, I'll to refactor to this style.

Reply:
TIL. Thank you for the feedback, I'll to refactor to this style.

Reply:
Ah right, I started out with this:

```elixir
  defp fetched?(result, batch_key, item_key) do
    match?(%{^batch_key => {:ok, %{^item_key => _value}}}, result)
  end
```

which worked and while writing the comment I thought, why not pattern match directly in the function's arguments, but didn't test that.

Reply:
Happy to merge after pattern matching tweaks are in.

Reply:
It should be ok now. :)